core/src/model.jl

```diff
@@ -188,10 +191,51 @@ function SciMLBase.successful_retcode(model::Model)::Bool
 end
 
 """
-    solve!(model::Model)::ODESolution
+    step!(model::Model, dt::Float64)::Model
+
+Take Model timesteps until `t + dt` is reached exactly.
+"""
+function SciMLBase.step!(model::Model, dt::Float64)::Model
+    (; config, integrator) = model
+    (; t) = integrator
+    # If we are at an allocation time, run allocation before the next physical
+    # layer timestep. This allows allocation over period (t, t + dt) to use variables
+    # set over BMI at time t before calling this function.
+    # Also, don't run allocation at t = 0 since there are no flows yet (#1389).
+    ntimes = t / config.allocation.timestep
+    if ntimes > 0 && round(ntimes) ≈ ntimes
```

Comment:
I'd say the more common way to do this is

```julia
if t % config.allocation.timestep ≈ 0.0
   ... 
end
```

and the other check could just be `t > 0` if you actually need that.

Reply:
Although the comparison tolerances work out differently

Reply:
That would only work if the t is right after, but not right before, right?
```
julia> prevfloat(3.0) % 0.5
0.49999999999999956
```